distributed/client.py

```diff
@@ -3034,23 +3034,12 @@ def restart(self, **kwargs):
         """
         return self.sync(self._restart, **kwargs)
 
-    async def _upload_file(self, filename, raise_on_error=True):
-        with open(filename, "rb") as f:
-            data = f.read()
-        _, fn = os.path.split(filename)
-        d = await self.scheduler.broadcast(
-            msg={"op": "upload_file", "filename": fn, "data": to_serialize(data)}
+    async def _upload_file(self, filename):
+        await self._register_worker_plugin(
+            UploadFile(filename),
+            name=filename + str(uuid.uuid4()),
```

Comment:
I suspect that we no longer need the split `upload_file/_upload_file` here, and can simplify this by using `register_worker_plugin` directly

```python
def upload_file(self, filename):
    return self.register_worker_plugin(
        UploadFile(filename),
        name=filename + str(uuid.uuid4()),
    )
```

The background here is that the `_underscored_functions` are all async while the `normal_functions` use the `Client.sync` method, which determines if we're in the same thread as an event loop or not, and calls the `_underscored_function` directly if we are, returning a future that the calling code then awaits, or if we are in a different thread it triggers the `_underscored_function` to be called on the event loop, and then blocks until it is done.

In our case, this magic is already handled by `register_worker_plugin` and so we can probably rely on that method's use of the `sync`.  We probably don't need to do this dance again.  Hooray for basing methods on top of other methods.

Reply:
(This sync magic is historically annoying to first get used to, my apologies.  It *is* nice once you start blending async and sync code though)